### PR TITLE
ISPN-6217 Created Javadocs for remote client distribution

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/ISPN-6217

### CODE REVIEW

**Comment by danberindei** [comment]
There are some conflicts, I believe from Tristan's changes.

**Comment by slaskawi** [comment]
Rebased

File: javadoc-remote/pom.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+-&lt;script type='text/javascript'&gt;
+-var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
+-document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
+-&lt;/script&gt;
+-&lt;script type='text/javascript'&gt;
+-try {
+-var pageTracker = _gat._getTracker("UA-8601422-4");
+-pageTracker._trackPageview();
+-} catch(err) {}&lt;/script&gt;
+            </analytics.code>
```
Comment (tristantarrant) [inline javadoc-remote/pom.xml]: We need to update the analytics code to the new async stuff:
`<script>
     (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
                        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
                        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
                        })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
     
    ga('create', 'UA-8601422-1', 'auto');
    ga('send', 'pageview');
    </script>`

**Comment by slaskawi** [comment]
@tristantarrant Updated. The API docs pages look good in the browser but I have no idea how to verify if they work correctly.

**Comment by tristantarrant** [comment]
The release script needs to be changed to also upload the remote javadocs.
Changes will also be required on the infinispan.github.io side to link to the remote javadocs.

**Comment by slaskawi** [comment]
@tristantarrant This might be a little bit tricky...

Looking at https://github.com/infinispan/infinispan/blob/master/bin/release.py#L202, I see that we upload only one Javadocs package. The [Infinispan documentation page](http://infinispan.org/documentation) has one reference to the apidocs and I think two links (as for the remote package and one for everything else) would be a bit confusing.

So my proposition is to add remote javadocs to `-all` distribution and leave the documentation page with 1 javadocs link. 

WDYT?

**Comment by slaskawi** [comment]
Thumbs up means yes :) Let me refactor this...

**Comment by slaskawi** [comment]
@tristantarrant Updated and rebased. 

Unfortunately I had to introduce another module called `javadocs-all` with all embedded modules and Hot Rod client. I also couldn't use transitive dependencies in `javadocs-all` since the javadoc plugin does not support it...

**Comment by slaskawi** [comment]
Rebased and updated...

@tristantarrant May I ask you to take a look?

**Comment by tristantarrant** [comment]
Actually @slaskawi it would be better for the "all" package to be simply named "javadoc" as before: in this way you won't have to fix the site

**Comment by slaskawi** [comment]
@tristantarrant I analyzed the PR again and it looks like we need 3 JavaDoc packages:

- Embedded (only with core javadocs used in `minimal` distribution)
- Remote (for `remote` one)
- All (for `all` distribution and website)

I think it's beneficial to add those suffixes (even though it might slightly complicate packaging scripts). Does it make sense

**Comment by slaskawi** [comment]
I'm turning it back to "read for review" state.

**Comment by tristantarrant** [comment]
@slaskawi I think we can safely avoid deploying the javadocs artifacts to Maven: they are only needed for the distribution packages.

**Comment by slaskawi** [comment]
@tristantarrant Fixed. Ready for review.

**Comment by tristantarrant** [comment]
Can you please move javadoc-[all|embeded|remote] as children of the javadoc directory ?

**Comment by slaskawi** [comment]
Of course... on my way...

**Comment by slaskawi** [comment]
@tristantarrant Moved javadocs to their own directory.

**Comment by tristantarrant** [comment]
Pushed, thanks

**Comment by slaskawi** [comment]
Awesome, thanks Tristan!

**Pull request closed without merging**